src/Microsoft.AspNet.OData.Shared/PerRouteContainerBase.cs

```diff
@@ -51,26 +54,75 @@ public IServiceProvider CreateODataRootContainer(Action<IContainerBuilder> confi
             return rootContainer;
         }
 
+        /// <summary>
+        /// Check if the root container for a given route name exists.
+        /// </summary>
+        /// <param name="routeName">The route name.</param>
+        /// <returns>true if root container for the route name exists, false otherwise.</returns>
+        public bool HasODataRootContainer(string routeName)
+        {
+            IServiceProvider rootContainer = this.GetContainer(routeName);
+            return rootContainer != null;
+        }
+
         /// <summary>
         /// Get the root container for a given route name.
         /// </summary>
         /// <param name="routeName">The route name.</param>
         /// <returns>The root container for the route name.</returns>
-        public abstract IServiceProvider GetODataRootContainer(string routeName);
+        /// <remarks>
+        /// This function will throw an exception if no container is found
+        /// in order to localize the failure and provide a consistent error
+        /// message. Use <see cref="HasODataRootContainer"/> to test of a container
+        /// exists without throwing an exception.
+        /// </remarks>
+        public IServiceProvider GetODataRootContainer(string routeName)
+        {
+            IServiceProvider rootContainer = this.GetContainer(routeName);
+            if (rootContainer == null)
+            {
+                if (String.IsNullOrEmpty(routeName))
+                {
+                    throw Error.InvalidOperation(SRResources.MissingNonODataContainer, routeName);
```

Comment:
I think this is right; when the routeName is null or empty, then we should use the "NonOData" container error message. When there is a route name, it's likely a bug because the route container should have been created in order to even get the route name included as part of the request.

Reply:
You're right, I jumped the gun by saying that they should be swapped. It look like SRResources.MissingNonODataContainer does not take a parameter for the error string though, and that the SRResources.MissingODataContainer one does. I think that's what made me think they were in swapped positions at first.

Reply:
Yep, the ", routeName" should be on the second exception, i.e. where routeName != null. Thanks!